src/kernel/bitmap.zig

```diff
@@ -1,205 +1,31 @@
-                    }
-                }
-                // Avoiding overflow by checking if bit is less than the max - 1
-                if (bit < NUM_ENTRIES - 1) {
-                    bit += 1;
-                } else {
-                    break;
-                }
-            }
-
-            if (count == num) {
-                if (start) |start_entry| {
-                    var i: IndexType = 0;
-                    while (i < num) : (i += 1) {
-                        self.setEntry(start_entry + i);
-                    }
-                    return start_entry;
-                }
-            }
-            return null;
-        }
-
-        ///
-        /// Set the first free entry within the bitmaps as occupied.
-        ///
-        /// Return: ?IndexType.
-        ///     The index within all bitmaps that was set or null if there wasn't one free.
-        ///     0 .. NUM_ENTRIES - 1 if in the first bitmap, NUM_ENTRIES .. NUM_ENTRIES * 2 - 1 if in the second etc.
-        ///
-        pub fn setFirstFree(self: *Self) ?IndexType {
-            if (self.num_free_entries == 0 or self.bitmap == BITMAP_FULL) {
-                std.debug.assert(self.num_free_entries == 0 and self.bitmap == BITMAP_FULL);
-                return null;
-            }
-            const bit = @truncate(IndexType, @ctz(BitmapType, ~self.bitmap));
-            self.setEntry(bit);
-            return bit;
-        }
-
-        ///
-        /// Check if an entry is set.
-        ///
-        /// Arguments:
-        ///     IN self: *const Self - The bitmap to check.
-        ///     IN idx: usize - The entry to check.
-        ///
-        /// Return: bool.
-        ///     True if the entry is set, else false.
-        ///
-        pub fn isSet(self: *const Self, idx: IndexType) bool {
-            return (self.bitmap & indexToBit(idx)) != 0;
-        }
-    };
-}
+/// The possible errors thrown by bitmap functions
+pub const BitmapError = error{
+    /// The address given was outside the region covered by a bitmap
+    OutOfBounds,
+};
 
 ///
 /// A bitmap that uses a specific type to store the entries.
+/// The bitmap can either be statically or dynamically allocated.
+/// Statically allocated bitmaps are allocated at compile time and therefore must know the number of entries at compile time.
+/// Dynamically allocated bitmaps do not need to know the number of entries until being initialised, but do need an allocator.
 ///
 /// Arguments:
+///     IN static: comptime bool - Whether this bitmap is statically or dynamically allocated
 ///     IN BitmapType: type - The integer type to use to store entries.
+///     IN num_entries: usize or ?usize - The number of entries if static, else ignored (can be null)
 ///
 /// Return: type.
 ///     The bitmap type created.
 ///
-pub fn Bitmap(comptime BitmapType: type) type {
+pub fn Bitmap(comptime static: bool, comptime BitmapType: type, comptime num_entries: if (static) usize else ?usize) type {
```

Comment:
Could you comptime check on whether `num_entries` is `null` and if null, then is static, else is not. So can use the 'null capture if' statement: `if (num_entries) |blah| {static} else {alloc}`. Then this would remove the `static` parameter.

Reply:
That's much nicer :)